### PR TITLE
fix: store readme when publishing locally

### LINKED ISSUE #1905: README published by yarn 2 is not correctly shown on verdaccio
⚠️⚠️⚠️⚠️ Workaround available, [scroll down](https://github.com/verdaccio/verdaccio/issues/1905#issuecomment-1057700082) ⬇️  ⚠️⚠️⚠️⚠️

**Describe the bug**

README is not correctly shown on verdaccio using `yarn npm publish`

**To Reproduce**

```sh
mkdir yarn-verdaccio-readme
cd yarn-verdaccio-readme
yarn set version berry
echo '
npmPublishRegistry: http://localhost:4873
unsafeHttpWhitelist: [localhost]
' >> .yarnrc.yml
echo '
{
  "name": "yarn-verdaccio-readme",
  "version": "0.1.0"
}
' > package.json
echo '# yarn-verdaccio-readme' > README.md
yarn
docker run --rm -p 4873:4873 -d verdaccio/verdaccio:4.8.1
echo 'input some random username/password'
yarn npm login --publish
yarn npm publish
yarn npm logout --publish
echo 'open http://localhost:4873/-/web/detail/yarn-verdaccio-readme then you will see "ERROR: No README data found!".'
```

**Expected behavior**

README should be correctly shown on verdaccio

**Screenshots**

```
$ yarn npm publish
➤ YN0000: README.md
➤ YN0000: package.json
➤ YN0000: Package archive published
➤ YN0000: Done in 0.91s
```

image

**Configuration File (cat ~/.config/verdaccio/config.yaml)**

Default config in the `verdaccio/verdaccio:4.8.1` docker image

**Environment information**

Server: the `verdaccio/verdaccio:4.8.1` docker image
Client: Yarn 2.1.1

**Additional context**

- `README.md` is correctly uploaded
- the issue seems to be that `verdaccio` depends on the `readme` field in the publish body to show the README content
- the yarn 2 client does not provide the `readme` field in the publish body while `npm` did provide it, but the NPM registry seems to work fine without that field, for example [the yarn 2 published `@yarnpkg/libzip` package](https://www.npmjs.com/package/@yarnpkg/libzip) shows the README correctly.
- the corresponding yarn 2 issue: https://github.com/yarnpkg/berry/issues/1702

### PR DESCRIPTION
**Pull request by mbtools**
When publishing to verdaccio, the readme data is not stored correctly in some cases (although it's included in the publish request). As a result, the readme is not visible in the web interface:

image

This fix and enhanced tests clarify how verdaccio stores readme data (currently):

- Depending on the used package manager, `publish` might include readme data at root level (`manifest.readme`) or on version level (`manifest.versions[version].readme`) or both
- Verdaccio will keep only the readme data of the **last published version** and store it at root level

Closes #1905 (and some discussion items)

PS: Renamed variable `manifest` to `localManifest` to avoid confusion with function parameter that had the same name.

### CODE REVIEW

**Comment by mbtools** [comment]
mmm, I don't see how the failing tests are related to the change 🤷

**Comment by juanpicado** [comment]
> mmm, I don't see how the failing tests are related to the change 🤷 

I will check when I have some time, there is a flaky test I've been pursuing and I could not get rid of it

**Comment by juanpicado** [comment]
> mmm, I don't see how the failing tests are related to the change 🤷

I might have improved the situation, please take latest from master branch

File: packages/store/src/storage.ts [hunk 1]
```diff
@@ -1127,9 +1127,12 @@ class Storage {
 
     // 1. after tarball has been successfully uploaded, we update the version
     try {
-      // TODO: review why do this
-      versions[versionToPublish].readme =
-        _.isNil(manifest.readme) === false ? String(manifest.readme) : '';
+      // Older package managers like npm6 do not send readme content as part of version but include it on root level
```
Comment (juanpicado) [inline packages/store/src/storage.ts:1130]: that's actually true, I had this logic but maybe somewhere but not here, thanks.

**Review by juanpicado — APPROVED** [review]
LGTM thanks :)

**Comment by juanpicado** [comment]
@mbtools thanks for the support, you may have noticed :-) small upgrade in your profile

**Comment by juanpicado** [comment]
some tests fails here `packages/web/test/api.readme.test.ts` I haven't checked but either the setup is wrong or the patch might be causing side effects, let me know if I can help here.

**Comment by mbtools** [comment]
Thanks, Juan. Nice to be part of the team 😃 I will have a look

**Pull request merged**